cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -517,36 +521,52 @@ public void upgradeApplications(FullHttpRequest request, HttpResponder responder
         jsonWriter.beginArray();
+        List<Future<?>> futureList = new ArrayList<>();
+        final ReentrantLock lock = new ReentrantLock();
+        ThreadPoolExecutor executor = (ThreadPoolExecutor) Executors.newFixedThreadPool(10);
+
         for (ApplicationId appId : appIds) {
-          ApplicationUpdateDetail updateDetail;
-          try {
-            applicationLifecycleService.upgradeApplication(appId, allowedArtifactScopes, allowSnapshot);
-            updateDetail = new ApplicationUpdateDetail(appId);
-          } catch (UnsupportedOperationException e) {
-            String errorMessage = String.format("Application %s does not support upgrade.", appId);
-            updateDetail = new ApplicationUpdateDetail(appId, new NotImplementedException(errorMessage));
-          } catch (InvalidArtifactException | NotFoundException e) {
-            updateDetail = new ApplicationUpdateDetail(appId, e);
-          } catch (Exception e) {
-            updateDetail =
+          Future<?> future = executor.submit(() -> {
+            ApplicationUpdateDetail updateDetail;
+            try {
+              applicationLifecycleService.upgradeApplication(appId, allowedArtifactScopes, allowSnapshot);
+              updateDetail = new ApplicationUpdateDetail(appId);
+            } catch (UnsupportedOperationException e) {
+              String errorMessage = String.format("Application %s does not support upgrade.", appId);
+              updateDetail = new ApplicationUpdateDetail(appId, new NotImplementedException(errorMessage));
+            } catch (InvalidArtifactException | NotFoundException e) {
+              updateDetail = new ApplicationUpdateDetail(appId, e);
+            } catch (Exception e) {
+              updateDetail =
                 new ApplicationUpdateDetail(appId, new ServiceException("Upgrade failed due to internal error.", e,
-                                            HttpResponseStatus.INTERNAL_SERVER_ERROR));
-            LOG.error("Application upgrade failed with exception", e);
-          }
-          GSON.toJson(updateDetail, ApplicationUpdateDetail.class, jsonWriter);
-          jsonWriter.flush();
-          chunkResponder.sendChunk(Unpooled.wrappedBuffer(outputStream.toByteArray()));
-          outputStream.reset();
-          chunkResponder.flush();
+                                                                        HttpResponseStatus.INTERNAL_SERVER_ERROR));
+              LOG.error("Application upgrade failed with exception", e);
+            }
+
+            lock.lock();
+            try {
+              GSON.toJson(updateDetail, ApplicationUpdateDetail.class, jsonWriter);
+              jsonWriter.flush();
+              chunkResponder.sendChunk(Unpooled.wrappedBuffer(outputStream.toByteArray()));
+              outputStream.reset();
+              chunkResponder.flush();
+            } catch (Exception e) {
+              throw new RuntimeException(e.getMessage());
```

Comment:
Always retain the original exception as the root cause when wrapping an exception (use `throw new RuntimeException(e);` instead)